File: js/leetcode/19-remove-nth-node-from-end-of-list.js
// https://leetcode.com/problems/remove-nth-node-from-end-of-list/

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function (head, n) {
  let counter = head;
  let count = 0;

  while (counter) {
    count++;
    counter = counter.next;
  }

  let k = count - n;
  let index = 0;
  let current = head;
  let prev = null;
  while (current) {
    if (index === k) {
      // if this is the head
      if (!prev) {
        head = current.next;
      } else {
        prev.next = current.next;
        break;
      }
    }
    prev = current;
    current = current.next;
    index++;
  }
  return head;
};

File: python/leetcode/19-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:

        current = head
        size = 0

        while current:
            size += 1
            if current.next is None:
                break
            current = current.next

        target = size - n
        n2 = 0
        current, previous = head, None
        while current:
            if n2 == target:
                if current == head:
                    head = current.next
                else:
                    previous.next = current.next
                break
            else:
                previous = current
                current = current.next
                n2 += 1

        return head
